// @simplr-wc/button/lib/button.js
import { __decorate } from "tslib";
import { html } from 'lit-html';
import { SimplrComponentBase, CustomElement, Property, css } from '@simplr-wc/core';
let SimplrButton = class SimplrButton extends SimplrComponentBase {
    constructor() {
        super();
        // Button status
        this.disabled = false;
        // Button style
        this.outlined = false;
        this.contained = false;
        this.elevated = false;
        // Button types
        this.primary = false;
        this.secondary = false;
        this.success = false;
        // Button props
        this.type = 'button';
        this.attachShadow({ mode: 'open' });
    }
    connectedCallback() {
        this.tabIndex = 0;
        this.createButton();
        this.addEventListeners();
    }
    updated() {
        this.updateButtonAttributes();
    }
    addEventListeners() {
        this.addEventListener('keyup', this.handleKeyboardEvent.bind(this));
        this.addEventListener('click', e => {
            // Don't bubble the click event from this one. Bubble it from the actual button
            if (e.target === this) {
                e.stopPropagation();
                this.doClick();
            }
        }, true);
    }
    handleKeyboardEvent(e) {
        if (e.key === ' ' || e.key === 'Enter') {
            this.doClick();
        }
    }
    doClick() {
        this.buttonElem?.click();
    }
    createButton() {
        this.buttonElem = document.createElement('button');
        this.buttonElem.tabIndex = -1;
        this.appendChild(this.buttonElem);
        this.updateButtonAttributes();
    }
    updateButtonAttributes() {
        if (this.buttonElem) {
            this.buttonElem.type = this.type;
        }
    }
    attributeChangedCallback(name, oldValue, newValue) {
        if (oldValue === newValue)
            return;
        this[name] = newValue === '' ? true : newValue;
    }
    static get observedAttributes() {
        return ['disabled', 'type', 'primary', 'secondary', 'success', 'outlined', 'contained', 'elevated'];
    }
    get html() {
        return html `<slot></slot>`;
    }
    get css() {
        return css `
            :host {
                --primary-color: #0087d7;
                --secondary-color: #f94416;
                --success-color: #41d888;

                --main-color: transparent;
                --text-color: var(--main-color);
                --ripple-color: var(--main-color);

                --size: 16px;

                position: relative;
                display: flex;
                align-items: center;
                text-align: center;
                font-size: var(--size);

                width: fit-content;
                cursor: pointer;
                line-height: 36px;
                padding: 0 16px;
                border-radius: 4px;
                outline: none;

                transition: 200ms ease-in-out;
                transition-property: background;

                background-color: transparent;
                color: var(--text-color);

                overflow: hidden;
            }

            :host([contained]) {
                background-color: var(--main-color);
                --text-color: #fff;
                --ripple-color: var(--text-color);
                font-weight: 700;
            }

            :host([outlined]) {
                border: 2px solid var(--main-color);
                --ripple-color: var(--text-color);
            }

            :host([elevated]) {
                --text-color: #fff;
                --ripple-color: var(--text-color);
                background-color: var(--main-color);
                font-weight: 700;
                box-shadow: 0px 3px 3px -2px rgba(0, 0, 0, 0.2), 0px 2px 2px 0px rgba(0, 0, 0, 0.14),
                    0px 1px 8px 0px rgba(0, 0, 0, 0.12);
            }

            :host([primary]) {
                --main-color: var(--primary-color);
            }
            :host([secondary]) {
                --main-color: var(--secondary-color);
            }
            :host([success]) {
                --main-color: var(--success-color);
            }

            :host(:focus) {
            }

            :host::after {
                content: '';
                position: absolute;
                top: 0;
                left: 0;
                width: 100%;
                height: 100%;
                background: var(--ripple-color);
                opacity: 0;
                transition: 200ms ease-in-out;
                pointer-events: none;
            }

            :host(:hover)::after,
            :host(:focus-visible)::after {
                opacity: 0.3;
            }

            :host::before {
                content: '';
                position: absolute;
                left: 0;
                top: 0;
                bottom: 0;
                right: 0;
                width: 2rem;
                height: 2rem;
                border-radius: 50%;
                margin: auto;
                background: var(--ripple-color);
                opacity: 0;
                transform: scale(8);
                transition: 600ms ease-in-out;
                pointer-events: none;
            }

            :host(:active)::before {
                opacity: 0.5;
                transform: scale(0);
                transition: 0ms;
            }

            ::slotted(svg) {
                fill: var(--text-color);
                height: var(--size);
                width: var(--size);
                padding-right: calc(var(--size) * 0.7);
            }

            ::slotted(button) {
                position: absolute;
                top: 0;
                left: 0;
                height: 0;
                width: 0;
                padding: 0;
                border: 0;
            }
        `;
    }
};
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "disabled", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "outlined", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "contained", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "elevated", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "primary", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "secondary", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "success", void 0);
__decorate([
    Property({ reflect: true })
], SimplrButton.prototype, "type", void 0);
__decorate([
    Property({})
], SimplrButton.prototype, "buttonElem", void 0);
SimplrButton = __decorate([
    CustomElement('simplr-button')
], SimplrButton);
export default SimplrButton;
